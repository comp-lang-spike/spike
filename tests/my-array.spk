#!/usr/bin/env spike

class MyArray : Array {
    obj label;
    
    label { return label; }
    label = aLabel { label = aLabel; }
    
    self[i] {
        obj oldValue;
        
        oldValue = super[i];
        printf("MyArray: self[%d] == %s\n", i, oldValue.printString);
        return oldValue;
    }
    self[i] = newValue {
        printf("MyArray: self[%d] = %s\n", i, newValue.printString);
        return super[i] = newValue;
    }
}

main(argv) {
    obj a, i;
    
    a = MyArray new: 4;
    a.label = "tomato";
    printf("The array size is %d.\n", a.size);
    for (i = 0; i < 4; ++i) {
        a[i] = i * 2;
    }
    for (i = 0; i < 4; ++i) {
        printf("%s\n", a[i].printString);
    }
    printf("The label is '%s'.\n", a.label);
    return 0;
}
