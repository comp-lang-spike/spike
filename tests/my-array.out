The array size is 4.
MyArray: self[0] = 0
MyArray: self[1] = 2
MyArray: self[2] = 4
MyArray: self[3] = 6
MyArray: self[0] == 0
0
MyArray: self[1] == 2
2
MyArray: self[2] == 4
4
MyArray: self[3] == 6
6
The label is 'tomato'.
